modules/javafx.graphics/src/main/java/com/sun/javafx/scene/layout/HeaderButtonBehavior.java

```diff
@@ -0,0 +1,115 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.javafx.scene.layout;
+
+import com.sun.javafx.PlatformUtil;
+import javafx.beans.value.ObservableValue;
+import javafx.event.EventHandler;
+import javafx.scene.Node;
+import javafx.scene.Scene;
+import javafx.scene.input.MouseEvent;
+import javafx.scene.layout.HeaderButtonType;
+import javafx.stage.Modality;
+import javafx.stage.Stage;
+import javafx.util.Subscription;
+import java.util.Objects;
+import java.util.Optional;
+
+public final class HeaderButtonBehavior implements EventHandler<MouseEvent> {
+
+    private final Node node;
+    private final HeaderButtonType type;
+    private final Subscription subscription;
+
+    public HeaderButtonBehavior(Node node, HeaderButtonType type) {
+        this.node = Objects.requireNonNull(node);
+        this.type = Objects.requireNonNull(type);
+
+        ObservableValue<Stage> stage = node.sceneProperty()
+            .flatMap(Scene::windowProperty)
+            .map(w -> w instanceof Stage s ? s : null);
+
+        subscription = Subscription.combine(
+            type == HeaderButtonType.MAXIMIZE
+                ? stage.flatMap(Stage::resizableProperty).subscribe(this::onResizableChanged)
+                : Subscription.EMPTY,
+            stage.flatMap(Stage::fullScreenProperty).subscribe(this::onFullScreenChanged),
+            () -> node.removeEventHandler(MouseEvent.MOUSE_RELEASED, this)
+        );
+
+        node.addEventHandler(MouseEvent.MOUSE_RELEASED, this);
+        node.setFocusTraversable(false);
+    }
+
+    public void dispose() {
+        subscription.unsubscribe();
+    }
+
+    @Override
+    public void handle(MouseEvent event) {
+        if (!node.getLayoutBounds().contains(event.getX(), event.getY())) {
+            return;
+        }
+
+        switch (type) {
+            case CLOSE -> getStage().ifPresent(Stage::close);
+            case ICONIFY -> getStage().ifPresent(stage -> stage.setIconified(true));
+            case MAXIMIZE -> getStage().ifPresent(stage -> {
+                    // On macOS, a non-modal window is put into full-screen mode when the maximize button is clicked,
+                    // but enlarged to cover the desktop when the option key is pressed at the same time.
+                    if (PlatformUtil.isMac() && stage.getModality() == Modality.NONE && !event.isAltDown()) {
+                        stage.setFullScreen(!stage.isFullScreen());
+                    } else {
+                        stage.setMaximized(!stage.isMaximized());
+                    }
+                });
```

Comment:
I think it might be better to add a default case here, _in case this enum evolves_.

Reply:
Why? This isn't some external enum that might evolve out from under this code, but is part of the same feature. If the enum evolves then the resulting compiler error is a _good_ thing as it alerts whoever added the new enum that they didn't finish implementing it.

A default makes sense when the enum might evolve separately from the use of the enum.